### PR TITLE
[terra-arrange][terra-section-header] update to utilize semantic html elements

### PR DESCRIPTION
**Pull request by KV106606Viswanath**
### Summary
Currently Toggle Section Header wraps a H2 with a button which might not be valid HTML. While the screen reader does read and interpret code correctly, the invalid code could fail for other browsers or assistive technologies. Added valid html to avoid these failures.

**What was changed:**
Updated HTML code div elements with button and span tags to make it valid html.

**Why it was changed:**
To avoid failures of invalid html on other browsers or assistive technologies.

**This PR resolves:**

UXPLATFORM-9093

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/terra-section-header/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,13 @@
 
 ## Unreleased
 
+* Changed
+  * Updated div element with span tag element.
+  * Updated div element with button tag element.
```
Comment (sdadn) [inline packages/terra-section-header/CHANGELOG.md:7]:
```suggestion
  * Updated SectionHeader to use `span` & `button` elements instead of `div`.
```

**Review by sdadn — CHANGES_REQUESTED** [review]
You may need to update jest snapshots as well.

File: packages/terra-arrange/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Changed
+  * Updated div elements with span tag elements.
```
[the remark this answers is not in the document]
Comment (KV106606Viswanath) [reply packages/terra-arrange/CHANGELOG.md]: Updated.

File: packages/terra-section-header/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,13 @@
...
+
+* Added
+  * Added a function "ArrangeComponent" to use Arrange component.
```
[the remark this answers is not in the document]
Comment (KV106606Viswanath) [reply packages/terra-section-header/CHANGELOG.md]: Removed these 2 lines.

File: packages/terra-section-header/CHANGELOG.md (hunk 1 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/CHANGELOG.md]: Updated.

File: packages/terra-arrange/src/Arrange.module.scss [hunk 4]
```diff
@@ -37,3 +37,16 @@
     align-self: flex-start;
   }
 }
+
+.fit-block {
+  .icon-wrapper {
+    display: inline-flex;
+    font-size: 8px;
```
Comment (supreethmr) [inline packages/terra-arrange/src/Arrange.module.scss:44]: Are these style values provided by UX..??

File: packages/terra-arrange/src/Arrange.jsx [hunk 5]
```diff
@@ -75,17 +75,17 @@ const Arrange = ({
...
   return (
-    <div {...customProps} className={cx('arrange', customProps.className)}>
-      <div {...fitStartProps} className={cx('fit', align || alignFitStart, fitStartProps.className)}>
+    <span {...customProps} className={cx('arrange', customProps.className)}>
+      <span {...fitStartProps} className={cx('fit', align || alignFitStart, fitStartProps.className || 'fit-block')}>
         {fitStart}
-      </div>
-      <div {...fillProps} className={cx('fill', align || alignFill, fillProps.className)}>
+      </span>
+      <span {...fillProps} className={cx('fill', align || alignFill, fillProps.className || 'fill-block')}>
```
Comment (supreethmr) [inline packages/terra-arrange/src/Arrange.jsx:82]:
```suggestion
      <span {...fillProps} className={cx('fill', align || alignFill, fillProps.className, 'fill-block')}>
```
why are using || here with fillProps.classname ?

File: packages/terra-arrange/src/Arrange.module.scss [hunk 6]
```diff
@@ -37,3 +37,16 @@
     align-self: flex-start;
   }
 }
+
+.fit-block {
+  .icon-wrapper {
```
Comment (supreethmr) [inline packages/terra-arrange/src/Arrange.module.scss:42]: Is this class intended for terra-arrange ? there are no reference in Arrange component for class `icon-wrapper` !!

File: packages/terra-arrange/src/Arrange.module.scss (hunk 4 above)
Comment (KV106606Viswanath) [reply packages/terra-arrange/src/Arrange.module.scss:44]: No, these are not provided by UX team. I have added them to match the existing look of UI. If required i will check with UX team.

File: packages/terra-arrange/src/Arrange.module.scss (hunk 6 above)
Comment (KV106606Viswanath) [reply packages/terra-arrange/src/Arrange.module.scss]: This class has been added in section header component, i have moved these styles into section header module scss file.

File: packages/terra-arrange/src/Arrange.jsx (hunk 5 above)
Comment (KV106606Viswanath) [reply packages/terra-arrange/src/Arrange.jsx]: Removed that || operator as it's not required here.

File: packages/terra-arrange/src/Arrange.module.scss (hunk 4 above)
Comment (supreethmr) [reply packages/terra-arrange/src/Arrange.module.scss]: yes we need UX review for the above style changes !!

File: packages/terra-core-docs/CHANGELOG.md
Comment (sdadn) [inline packages/terra-core-docs/CHANGELOG.md:1]: Can you rebase with main so that this file is up to date?

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 7]
```diff
@@ -153,17 +153,28 @@ class SectionHeader extends React.Component {
...
 
-    const buttonAttributes = (onClick) ? { role: 'button', 'aria-expanded': isOpen, 'aria-label': headerText } : undefined;
+    // Function to use Arrange component
+    const ArrangeComponent = () => (
+      <Arrange
+        fitStart={onClick && accordionIcon}
+        fill={<span aria-hidden={(onClick !== undefined)} className={cx('title', 'title-block')}>{headerText}</span>}
+        className={cx('title-arrange')}
+      />
+    );
```
Comment (sdadn) [inline packages/terra-section-header/src/SectionHeader.jsx:163]: This can be a variable instead of a function.

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 8]
```diff
@@ -153,17 +153,28 @@ class SectionHeader extends React.Component {
...
+
+    const buttonAttributes = (onClick) ? { 'aria-expanded': isOpen, 'aria-label': headerText } : undefined;
```
Comment (sycombs) [inline packages/terra-section-header/src/SectionHeader.jsx:165]: Are these still needed? Should `aria-expanded': isOpen` be set on the `<button>` and `'aria-label': headerText` be set on the `<span>`?

**Comment by sycombs** [comment]
Why is the text color different?
image

https://github.com/cerner/terra-core/blob/0a838433201102146b09c06c762cfe6218f55f11/packages/terra-section-header/tests/wdio/__snapshots__/reference/clinical-lowlight-theme/en/chrome_huge/section-header-spec/long_title_accordion_default.png

File: packages/terra-section-header/src/SectionHeader.jsx (hunk 8 above)
Comment (sycombs) [reply packages/terra-section-header/src/SectionHeader.jsx:165]: I realized I misread this. I don't think we need the ternary anymore since we can set `{ 'aria-expanded': isOpen, 'aria-label': headerText }` directly on the `<button>`.

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 9]
```diff
@@ -153,17 +153,28 @@ class SectionHeader extends React.Component {
...
-          />
-        </div>
+        {onClick && accordionIcon ? (
+          <button {...buttonAttributes} type="button" tabIndex="-1" className={cx('arrange-wrapper toggle-button')}>
+            <ArrangeComponent />
+          </button>
+        ) : (
+          <span tabIndex="-1" className={cx('arrange-wrapper')}>
+            <ArrangeComponent />
+          </span>
```
Comment (sdadn) [inline packages/terra-section-header/src/SectionHeader.jsx:176]: Both of these seem very similar so instead of defining these twice, we can do something like:

```js
const ArrangeWrapperTag = onClick ? "button" : "span" ;
const type = onClick ? "button" : null;
// ... and so on for the other attributes

<ArrangeWrapperTag  (attributes) >
  <ArrangeComponent />
</ArrangeWrapperTag


```

Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.jsx]: Updated.

File: packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx [hunk 10]
```diff
@@ -10,7 +10,7 @@ import ToggleSectionHeaderPropsTable from 'terra-toggle-section-header/src/Toggl
 
 # Terra Toggle Section Header
 
-Toggle section header component that transitions content in and out with a click.
+Toggle section header component that transitions content in and out with a click. For accessibility best practices, it is recommended that consumers should always use only one h1 tag per page or view. The one h1 tag should be the page title. A section header should never be a heading level 1.
```
Comment (sycombs) [inline packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx:13]: Can we move this accessibility info to its own section? Example: https://engineering.cerner.com/terra-ui/components/cerner-terra-core-docs/alert/about#accessibility

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 11]
```diff
@@ -153,17 +153,22 @@ class SectionHeader extends React.Component {
     /* eslint-disable jsx-a11y/click-events-have-key-events */
     /* eslint-disable jsx-a11y/no-static-element-interactions */
 
-    const buttonAttributes = (onClick) ? { role: 'button', 'aria-expanded': isOpen, 'aria-label': headerText } : undefined;
+    const buttonAttributes = (onClick) ? {
+      type: 'button', tabIndex: '-1', 'aria-expanded': isOpen, 'aria-label': headerText,
+    } : undefined;
+    const spanAttributes = { tabIndex: '-1' };
```
Comment (sycombs) [inline packages/terra-section-header/src/SectionHeader.jsx:159]: Right now it is possible (if onClick=true and accordionIcon=false) to have  an `ArrangeWrapperTag` that is a `span` with `buttonAttributes` and `spanAttributes`. Is this intended or should only these two cases:

1. `ArrangeWrapperTag` is a `button` with `buttonAttributes`
2. `ArrangeWrapperTag` is a `span` with `spanAttributes`

be allowed?

**Review by sycombs — CHANGES_REQUESTED** [review]

File: packages/terra-core-docs/CHANGELOG.md [hunk 12]
```diff
@@ -2,15 +2,29 @@
...
+
+## 1.29.0 - (June 28, 2023)
+
 * Changed
   * Updated `terra-status-view - Change Variant` example component to use the apply button to change the selected variant.
-  
+
+* Added
+  * Added documentation updates for `Single Select` in `terra-form-select`.
+
```
[the remark this answers is not in the document]
Comment (KV106606Viswanath) [reply packages/terra-core-docs/CHANGELOG.md:21]: We have new entries on main branch. Updated those entries here to make this file up to date as per saad's comment.

File: packages/terra-core-docs/CHANGELOG.md [hunk 13]
```diff
@@ -2,15 +2,29 @@
...
 ## Unreleased
 
+* Changed
+  * Updated Arrange component example to use `span` element instead of `div`.
+
+* Added
+  * Added test example for controlled form native select component.
+
+* Added
+  * Added more realistic examples for terra-toggle-button.
```
[the remark this answers is not in the document]
Comment (KV106606Viswanath) [reply packages/terra-core-docs/CHANGELOG.md]: Updated

File: packages/terra-section-header/src/SectionHeader.jsx (hunk 11 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.jsx]: We have two section headers one is plain section header (without toggle button) which will render inside span tag and another one is section header with toggle button which will render with button tag. So by using the "onClick" and "accordionIcon" conditions we were rendering button and span elements. We have removed the variable "spanAttributes" and added the corresponding attribute "tabIndex: -1" to else condition of buttonAttributes variable. And also updated the variable name "buttonAtrributes" to "elementAttributes". We have added "ArrangeWrapperTag" which was suggested by saad on the previous comment. https://github.com/cerner/terra-core/pull/3831#discussion_r1256454357

File: packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx (hunk 10 above)
Comment (KV106606Viswanath) [reply packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx:13]: We have added this text in Toggle section header implementation page which was suggested by UX team as part of this PR. Could you please provide some clarification on this ?

Comment (sycombs) [reply packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx:13]: I only linked terra-alert as an example for the layout for the component doc. I was asking to create an "Accessibility" section within the Terra Toggle Section Header page and move this text into that section. The 'Terra Toggle Section Header' section (and corresponding sections in other components) seem like they are generally used for a basic description of the component and its typical use, not for prescribing implementation best practices.

**Comment by KV106606Viswanath** [comment]
> Why is the text color different? image
> 
> https://github.com/cerner/terra-core/blob/0a838433201102146b09c06c762cfe6218f55f11/packages/terra-section-header/tests/wdio/__snapshots__/reference/clinical-lowlight-theme/en/chrome_huge/section-header-spec/long_title_accordion_default.png



> Why is the text color different? image
> 
> https://github.com/cerner/terra-core/blob/0a838433201102146b09c06c762cfe6218f55f11/packages/terra-section-header/tests/wdio/__snapshots__/reference/clinical-lowlight-theme/en/chrome_huge/section-header-spec/long_title_accordion_default.png

Updated

File: packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx (hunk 10 above)
Comment (KV106606Viswanath) [reply packages/terra-core-docs/src/terra-dev-site/doc/toggle-section-header/About.1.doc.mdx]: Updated

File: packages/terra-core-docs/CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (sdadn) [reply packages/terra-core-docs/CHANGELOG.md:19]: Looks like this line is already in main; your PR shouldn't be updating it. Can you rebase with main again to see if it gets corrected?

File: packages/terra-core-docs/CHANGELOG.md [hunk 14]
```diff
@@ -2,15 +2,27 @@
...
+* Added
+  * Added documentation updates for `Single Select` in `terra-form-select`.
+
 ## 1.28.0 - (June 14, 2023)
 
 * Added
   * Added terra-scroll A11y tests.
   * Added an example in terra-content-container without interactive elements.
-
+  * Added examples for terra-content-container with dark colors.
```
Comment (sdadn) [inline packages/terra-core-docs/CHANGELOG.md:25]: Same as: https://github.com/cerner/terra-core/pull/3831/files#r1261701315

File: packages/terra-arrange/src/Arrange.module.scss (hunk 4 above)
Comment (KV106606Viswanath) [reply packages/terra-arrange/src/Arrange.module.scss]: Got +1 from UX

File: packages/terra-core-docs/CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (KV106606Viswanath) [reply packages/terra-core-docs/CHANGELOG.md]: Rebased with main and now the file is up to date.

File: packages/terra-core-docs/CHANGELOG.md (hunk 14 above)
Comment (KV106606Viswanath) [reply packages/terra-core-docs/CHANGELOG.md]: Rebased with main and now the file is up to date.

**Review by sdadn — APPROVED** [review]

**Review by sycombs — APPROVED** [review]

File: packages/terra-core-docs/src/terra-dev-site/doc/arrange/common/clinical-lowlight-theme/examplesetup.scss [hunk 15]
```diff
@@ -0,0 +1,11 @@
+// stylelint-disable scss/at-mixin-pattern
+@import '~terra-mixins/lib/Mixins';
+
+.clinical-lowlight-theme {
```
Comment (supreethmr) [inline packages/terra-core-docs/src/terra-dev-site/doc/arrange/common/clinical-lowlight-theme/examplesetup.scss:4]: why do we need lowlight-theme file here ..??

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 16]
```diff
@@ -153,17 +153,21 @@ class SectionHeader extends React.Component {
     /* eslint-disable jsx-a11y/click-events-have-key-events */
     /* eslint-disable jsx-a11y/no-static-element-interactions */
 
-    const buttonAttributes = (onClick) ? { role: 'button', 'aria-expanded': isOpen, 'aria-label': headerText } : undefined;
+    const elementAttributes = (onClick && accordionIcon) ? {
+      type: 'button', tabIndex: '-1', 'aria-expanded': isOpen, 'aria-label': headerText,
+    } : { tabIndex: '-1' };
```
Comment (supreethmr) [inline packages/terra-section-header/src/SectionHeader.jsx:158]:
```suggestion
const buttonAttributes = (onClick && accordionIcon) ? { role: 'button', 'aria-expanded': isOpen, 'aria-label': headerText } : undefined;
```
let's keep this attributes as button since these attributes are added only for button.

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 17]
```diff
@@ -153,17 +153,21 @@ class SectionHeader extends React.Component {
...
+    const ArrangeWrapperTag = onClick && accordionIcon ? 'button' : 'span';
```
Comment (supreethmr) [inline packages/terra-section-header/src/SectionHeader.jsx:159]: Nit:
```suggestion
    const ArrangeWrapper = onClick && accordionIcon ? 'button' : 'span';
```

File: packages/terra-arrange/src/Arrange.module.scss [hunk 18]
```diff
@@ -37,3 +37,9 @@
     align-self: flex-start;
   }
 }
+
+.fill-block {
+  .title-block {
```
Comment (supreethmr) [inline packages/terra-arrange/src/Arrange.module.scss:42]: this class seems to be required by terra-section-header and not required by terra-arrange. 
if css variable is added then it needs value on other themes as well !!

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 19]
```diff
@@ -153,17 +153,21 @@ class SectionHeader extends React.Component {
...
+    const elementClassNames = onClick && accordionIcon ? cx('arrange-wrapper toggle-button') : cx('arrange-wrapper');
```
Comment (supreethmr) [inline packages/terra-section-header/src/SectionHeader.jsx:160]:
```suggestion
    const buttonClassName = onClick && accordionIcon ? cx('toggle-button') : undefined;
```

File: packages/terra-section-header/src/SectionHeader.jsx [hunk 20]
```diff
@@ -153,17 +153,21 @@ class SectionHeader extends React.Component {
...
 
     return (
       <Element {...attributes} className={sectionHeaderClassNames} aria-label={!onClick ? headerText : undefined}>
-        <div {...buttonAttributes} tabIndex="-1" className={cx('arrange-wrapper')}>
+        <ArrangeWrapperTag {...elementAttributes} className={elementClassNames}>
```
Comment (supreethmr) [inline packages/terra-section-header/src/SectionHeader.jsx:164]:
```suggestion
        <ArrangeWrapperTag {...elementAttributes} className={cx('arrange-wrapper, buttonClassName)}>
```

File: packages/terra-section-header/src/SectionHeader.module.scss [hunk 21]
```diff
@@ -152,3 +153,19 @@
...
+  color: inherit;
+  cursor: pointer;
+  padding: 0;
+  text-align: left;
+}
+
+.fit-block {
+  .icon-wrapper {
+    display: inline-flex;
+    font-size: 8px;
```
Comment (supreethmr) [inline packages/terra-section-header/src/SectionHeader.module.scss:169]:
```suggestion
    font-size: 8px;
```
do we need font-size for icon ..??

File: packages/terra-core-docs/src/terra-dev-site/doc/arrange/common/clinical-lowlight-theme/examplesetup.scss (hunk 15 above)
Comment (KV106606Viswanath) [reply packages/terra-core-docs/src/terra-dev-site/doc/arrange/common/clinical-lowlight-theme/examplesetup.scss:4]: The interactive elements which i added like home, menu and settings button etc are in black colour. These buttons are not visible when i switched to clinical theme. For that i have added clinical theme scss file and added svg icons with white color.

File: packages/terra-section-header/src/SectionHeader.module.scss (hunk 21 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.module.scss]: Earlier i got some minor font difference on the icon but now it looks good. So I have removed this font size.

File: packages/terra-section-header/src/SectionHeader.jsx (hunk 20 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.jsx]: Updated

File: packages/terra-section-header/src/SectionHeader.jsx (hunk 19 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.jsx]: Updated

File: packages/terra-section-header/src/SectionHeader.jsx (hunk 17 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.jsx]: Updated

File: packages/terra-section-header/src/SectionHeader.jsx (hunk 16 above)
Comment (KV106606Viswanath) [reply packages/terra-section-header/src/SectionHeader.jsx]: Updated

File: packages/terra-core-docs/CHANGELOG.md [hunk 22]
```diff
@@ -2,6 +2,13 @@
...
 
+* Changed
+  * Updated Arrange component example to use `span` element instead of `div`.
+
+* Added
+  * Added realistic examples for `terra-arrange`.
+  * Added realistic examples for `terra-toggle-section-header`.
+
 * Added
   * Added email validation for `terra-form-field`.
```
Comment (sycombs) [inline packages/terra-core-docs/CHANGELOG.md:13]: Let's combine the Added sections: 
```suggestion
* Added
  * Added realistic examples for `terra-arrange`.
  * Added realistic examples for `terra-toggle-section-header`.
  * Added email validation for `terra-form-field`.
```

File: packages/terra-core-docs/CHANGELOG.md [hunk 23]
```diff
@@ -40,7 +47,6 @@
 * Added
   * Added terra-scroll A11y tests.
   * Added an example in terra-content-container without interactive elements.
-  * Added examples for terra-content-container with dark colors.
```
Comment (sycombs) [inline packages/terra-core-docs/CHANGELOG.md:43]: Why was this removed?

File: packages/terra-toggle-section-header/README.md [hunk 24]
```diff
@@ -4,7 +4,7 @@
 [![NPM version](https://badgen.net/npm/v/terra-toggle-section-header)](https://www.npmjs.org/package/terra-toggle-section-header)
 [![Build Status](https://badgen.net/travis/cerner/terra-core)](https://travis-ci.com/cerner/terra-core)
 
-Toggle section header component that transitions content in and out with a click.
+Toggle section header component that transitions content in and out with a click. For accessibility best practices, it is recommended that consumers should always use only one h1 tag per page or view. The one h1 tag should be the page title. A section header should never be a heading level 1.
```
Comment (sycombs) [inline packages/terra-toggle-section-header/README.md:7]: Since this is already in the terra-toggle-section-header `About` doc, can we remove it from the README? Consumers don't generally refer to the README for component guidance.

Comment (KV106606Viswanath) [reply packages/terra-toggle-section-header/README.md]: Removed.

File: packages/terra-core-docs/CHANGELOG.md (hunk 22 above)
Comment (KV106606Viswanath) [reply packages/terra-core-docs/CHANGELOG.md]: Updated.

File: packages/terra-core-docs/CHANGELOG.md (hunk 23 above)
Comment (KV106606Viswanath) [reply packages/terra-core-docs/CHANGELOG.md]: This entry was now Added.

**Review by supreethmr — APPROVED** [review]
